Extract Common Fields in WIT's to Base Type

alexeykazakov opened the pull request:
A new base type **system.planneritem** is created.
This base type has all common fields such as:
- system.remote_item_id
- system.title
- system.description
- system.state
- system.assignee
- system.creator

which are inherited via **extendedTypeName="system.planneritem"** by the default types when the types are created/updated:
- system.userstory
- system.valueproposition
- system.fundamental
- system.experience
- system.feature
- system.bug

Fixes #288

**TODO:**
- add some tests

alexeykazakov commented:
Added a commit which introduces **parentPath** to the WIT media type.
Please let me know if this all makes sense to you so I'll add some tests.

tsmaeder commented on migration/migration.go:
```diff
@@ -275,23 +255,64 @@ func createOrUpdateCommon(typeName string, ctx context.Context, witr models.Work
-		fmt.Println("Work item type exists, will update/overwrite the fields only ")
-		convertedFields, err := models.TEMPConvertFieldTypesToModel(workItemTypeFields)
+		log.Printf("Work item type %v exists, will update/overwrite the fields only and parentPath", typeName)
+		path := "/"
```
If possible, I would move this code to the WITRepository in the models package. I feel uncomfortable having the logic for updating the parent path spread around.

alexeykazakov replied:
WIT is immutable. We could introduce Update for the implementation in models.WorkIemRepository without exposing it to the interface in application.WorkItemRepository and work directly with the implementation here, but IMO it's safer to keep the code which updates WIT in migration.go. WDYT?

tsmaeder replied:
After some discussions with other people, I think we should leave this here for the moment. Desiging an API for updating work item types is a larger undertaking.

tsmaeder commented on design/media_types.go:
```diff
@@ -90,15 +90,18 @@ var workItemType = MediaType("application/vnd.workitemtype+json", func() {
 		Attribute("version")
 		Attribute("name")
 		Attribute("fields")
+		Attribute("parentPath")
```
I would not expose the parent path to clients until we have a need for it. So far inheritance is only know to clients when creating work item types, not when using them, so this would be a larger design change than you would think.

alexeykazakov replied:
I was in doubt about this as well. Not sure at all if it's a good idea to expose the parent path. But if we don't need it for clients why do we need it in the DB? I think some alignment here would be a good think.
But as I said before when updating the existing WIPs during migration I can workaround this inconsistency between model and DB working with DB parent-path fields directly and not using model.WIP if we really don't want to expose parentPath.

tsmaeder replied:
I put it in there because I thought it would be useful for type based queries (is X a descendent of Y).

alexeykazakov replied:
What such queries will give you if it's not available for clients?

tsmaeder commented on migration/migration.go:
```diff
@@ -206,52 +206,32 @@ func PopulateCommonTypes(ctx context.Context, db *gorm.DB, witr models.WorkItemT
 		return err
 	}
-	if err := createOrUpdateSystemFeature(ctx, witr, db); err != nil {
+	if err := createOrUpdateCommon(models.SystemExperience, ctx, witr, db); err != nil {
```
Perhaps "createOrUpdatePlannerItemType"? Because there are no function overloading in Go, I try to make function names more unique than in other languages.

tsmaeder commented on migration/migration.go:
```diff
@@ -275,23 +255,64 @@ func createOrUpdateCommon(typeName string, ctx context.Context, witr models.Work
+				return err
+			}
+			path = extendedWit.ParentPath
+			if path == "/" {
```
This needs a check for circularity.

alexeykazakov replied:
We don't have to if WIT is immutable. During migration we are building WIT relations from scratch.

tsmaeder replied:
True dat!

tsmaeder commented on migration/migration.go:
```diff
@@ -275,23 +255,64 @@ func createOrUpdateCommon(typeName string, ctx context.Context, witr models.Work
 		jsonArray, err := json.Marshal(convertedFields)
 		jsonString := string(jsonArray[:])
-		q := fmt.Sprintf(`UPDATE work_item_types SET fields='%s' where name = '%s'`, jsonString, typeName)
+		q := fmt.Sprintf(`UPDATE work_item_types SET fields='%s',parent_path='%s' where name = '%s'`, jsonString, path, typeName)
```
Why are we not using gorm to update the WIT here?

alexeykazakov replied:
WIT is immutable. There is no Update. We need to update it during migration only. At least for now.

tsmaeder replied:
I'm not talking about the repository, just gorm.DB.Save() instead of raw SQL. That will update the "updatedAt" timestamp, for example.

alexeykazakov replied:
Ah, OK. Sounds reasonable.

alexeykazakov replied:
We actually can't get rid of raw SQL UPDATE here completely if we don't want to expose parentPath attribute to WIT mediatype. So we can't update parentPath in DB using db.Save() because parentPath is not available in WIT.

alexeykazakov replied:
And one more side effect of not exposing parentPath is that this migrations updates parentPath correctly only if we create/update one level deep extensions. For example PlannerItem->Bug.
Relationships like PlannerItem->Issue->Bug or deeper won't be updated correctly in parenPath.

alexeykazakov commented:
Rebased, fixed conflicts and did some minor changes: Renamed CreateOrUpdateCommon() to CreateOrUpdatePlannerItemTypeExtention().

The biggest question is should we expose parentPath to clients or not. IMO we should expose it or remove it everywhere (it currently exists in model.WIT and DB only).

tsmaeder commented on design/media_types.go:
```diff
@@ -129,15 +129,18 @@ var workItemType = a.MediaType("application/vnd.workitemtype+json", func() {
 	a.Attribute("version", d.Integer, "Version for optimistic concurrency control")
 	a.Attribute("name", d.String, "User Readable Name of this item type")
 	a.Attribute("fields", a.HashOf(d.String, fieldDefinition), "Definitions of fields in this work item type")
+	a.Attribute("parentPath", d.String, "The names of the parent types separated with '/' or just '/' if no parents")
```
I really dont' think this should be part of this PR. We can discuss the utility of having this exposed, but in the least, it has nothing to do with the task at hand.

alexeykazakov commented:
I rebased the PR and removed parentPath exposure from the WIT mediatype.

alexeykazakov commented:
@tsmaeder OK. I found a way to load models.WorkItemType instead of app.WorkItemType so I can do db.Save() + update parentPath correctly for extensions like BaseType->SubType1->SubType2 . See the new commit.
But in the update almightly-ui I now see this new planeritem:

![screenshot from 2016-11-16 18-15-09](https://cloud.githubusercontent.com/assets/620087/20373785/13068c42-ac29-11e6-9308-f6b2e9cd963c.png)

I think we should just fix it on the UI side only and simply filter "unknown" types there. Or we could add some attribute to the model/DB and media type ("baseType" or "systemType" = true or something like that) to let UI hide such special types.
WDYT?

cc: @joshuawilson

tsmaeder commented:
@alexeykazakov @joshuawilson I think which work item types can be created will be part of the per project configuration eventually. As far as I'm concerned, the problem is not directly related with this PR and I would live with the problem for now.

joshuawilson commented:
The UI downloads the list and if we filter the list based on the names not to show we might as well hard code what we need to show. Since we are not going to do that, is there something else we could filter on? Like if it has a parent or something?

tsmaeder commented:
I am reluctant to introduce a mechanism that is not correct. In the future, the project will have a set of allowable types. But this is out of scope for the current PR, in my opinion. For the time being, I would just live with the extra entry in the list.

tsmaeder commented:
[test]

alexeykazakov commented:
> In the future, the project will have a set of allowable types.

Great. Then I agree that we don't need to introduce any new mechanisms to filter base types. 
The only think we could do now is to add some default icon for "unknown" types in UI. Because I see that UI uses specific icons for every type but **planneritem** doesn't have any. And it seems that the only way to change the type of the item is to click on the icon. So I can't change the type later if I used **planneritem** for my item. Or we can ignore it for now if you are planning to change how you update types of work items in UI and missing icon won't be a constraint anymore.

tsmaeder commented:
[test]

tsmaeder commented:
created https://github.com/almighty/almighty-core/issues/492 to track the work items type filtering for the UI.
